component/discovery/kubernetes_crds/configGenerator.go

```diff
@@ -0,0 +1,225 @@
+package kubernetes_crds
+
+// SEE https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/promcfg.go
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/pkg/util/k8sfs"
+	v1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	commonConfig "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	promk8s "github.com/prometheus/prometheus/discovery/kubernetes"
+	"github.com/prometheus/prometheus/model/relabel"
+)
+
+type configGenerator struct {
+	config   *Config
+	logger   log.Logger
+	secretfs *k8sfs.FS
+}
+
+// the k8s sd config is mostly dependent on our local config for accessing the kubernetes cluster.
+// if undefined it will default to an in-cluster config
+func (cg *configGenerator) generateK8SSDConfig(
+	namespaceSelector v1.NamespaceSelector,
+	namespace string,
+	role promk8s.Role,
+	attachMetadata *v1.AttachMetadata,
+) *promk8s.SDConfig {
+	cfg := &promk8s.SDConfig{
+		Role: role,
+	}
+	namespaces := cg.getNamespacesFromNamespaceSelector(namespaceSelector, namespace)
+	if len(namespaces) != 0 {
+		cfg.NamespaceDiscovery.Names = namespaces
+	}
+	if cg.config.KubeConfig != "" {
+		cfg.KubeConfig = cg.config.KubeConfig
+	}
+	if cg.config.ApiServerConfig != nil {
+		apiCfg := cg.config.ApiServerConfig
+		hCfg := apiCfg.HTTPClientConfig
+		cfg.APIServer = apiCfg.Host.Convert()
+
+		if hCfg.BasicAuth != nil {
+			cfg.HTTPClientConfig.BasicAuth = hCfg.BasicAuth.Convert()
+		}
+
+		if hCfg.BearerToken != "" {
+			cfg.HTTPClientConfig.BearerToken = commonConfig.Secret(hCfg.BearerToken)
+		}
+		if hCfg.BearerTokenFile != "" {
+			cfg.HTTPClientConfig.BearerTokenFile = hCfg.BearerTokenFile
+		}
+		cfg.HTTPClientConfig.TLSConfig = *hCfg.TLSConfig.Convert()
+		if hCfg.Authorization != nil {
+			if hCfg.Authorization.Type == "" {
+				hCfg.Authorization.Type = "Bearer"
+			}
+			cfg.HTTPClientConfig.Authorization = hCfg.Authorization.Convert()
+		}
+	}
+	if attachMetadata != nil {
+		cfg.AttachMetadata.Node = attachMetadata.Node
+	}
+	return cfg
+}
+
+func parseRegexp(s string, l log.Logger) relabel.Regexp {
+	r, err := relabel.NewRegexp(s)
+	if err != nil {
+		level.Error(l).Log("msg", "failed to parse regex", "err", err)
+	}
+	return r
```

Comment:
We shouldn't just be dropping the errors here. This should bubble up to the caller so we can set information in the DebugInfo to indicate there's a problem with the resource we discovered.